token/src/lib.rs

```diff
@@ -223,14 +224,59 @@ pub enum Error {
 
     /// Base64 decode error: {0}
     Base64Error(base64::DecodeError),
+
+    /// RSA Key error: {0}
+    RsaKeyError(rsa::pkcs1::Error),
+
+    /// Failure decoding the base64 layer of the base64 encoded PEM
+    Utf8Error(std::str::Utf8Error),
+}
+
+/// The supported JWT signature types.
+pub enum SignatureType {
+    HS256(HS256Key),
+    RS256(RS256KeyPair),
 }
 
 impl Token {
     /// Verifies and decodes a token.
-    pub fn from_jwt(token: &str, key: &HS256Key) -> Result<Self> {
-        key.verify_token(token, None)
-            .map_err(Error::TokenError)
-            .map(Token)
+    pub fn from_jwt(
+        token: &str,
+        signature_type: &SignatureType,
+        maybe_bound_issuer: &Option<String>,
+        maybe_bound_audiences: &Option<HashSet<String>>,
+    ) -> Result<Self> {
+        let opts = VerificationOptions {
+            reject_before: None,
+            accept_future: false,
+            required_subject: None,
+            required_key_id: None,
+            required_public_key: None,
+            required_nonce: None,
+            allowed_issuers: maybe_bound_issuer
+                .as_ref()
+                .map(|s| [s.to_owned()].into())
+                .to_owned(),
+            allowed_audiences: maybe_bound_audiences.to_owned(),
+            time_tolerance: None,
+            max_validity: None,
+            max_token_length: None,
+            max_header_length: None,
+        };
+
+        match signature_type {
+            SignatureType::HS256(key) => key
+                .verify_token(token, Some(opts))
+                .map_err(Error::TokenError)
+                .map(Token),
+            SignatureType::RS256(key) => {
+                let public_key = key.public_key();
```

Comment:
I think it should be possible to only provide the public key to atticd, without the private key. If the public key is always derived from the private key, that kind of voids the purpose of asymmetric JWTs. Also splitting this into public key for atticd and private key for atticadm means we can user other authorities like Keycloak without exposing their private key, as the signing would be done by Keycloak itself and atticd only needs to verify.

Reply:
We actually have an implementation of this, but it's not ready to PR yet (and I don't want to block this PR on that being ready).

Reply:
Considering that this PR is still pending two months later, maybe it's ready to include in this PR now? :)